### ISSUE TITLE
Core `lib/git.zsh` depends on `current_branch()` from git plugin

### ISSUE THREAD

**Issue by apjanke** [robbyrussell/oh-my-zsh#4085]
Some of the theme-related functions [defined in `lib/git.zsh`](https://github.com/robbyrussell/oh-my-zsh/blob/599831b85a5fd4564ab820530e2bb45f37ca3f99/lib/git.zsh#L70-L91) call `current_branch`. But `current_branch` is only defined in the `git` plugin. So if that plugin is not loaded, themes which use these functions will produce an error like `git_prompt_ahead:1: command not found: current_branch` each time the prompt is issued.

Functions affected:
- `git_prompt_ahead()`
- `git_prompt_behind()`
- `git_prompt_remote()`

Themes affected:
```
➜  .oh-my-zsh git:(master) egrep -r -l 'git_prompt_(ahead|behind|remote)' themes/*
themes/juanghurtado.zsh-theme
themes/sunrise.zsh-theme
themes/wedisagree.zsh-theme
➜  .oh-my-zsh git:(master) grep -rl current_branch themes/*
themes/eastwood.zsh-theme
themes/gallois.zsh-theme
themes/josh.zsh-theme
themes/juanghurtado.zsh-theme
themes/mortalscumbag.zsh-theme
themes/peepcode.zsh-theme
themes/sunrise.zsh-theme
```

I don't think the core stuff should depend on plugins, or at least, it should be an optional, declared dependency.

Workaround is to load the `git` plugin. Easy enough, but I don't think it should be required.

The `current_branch()` function definition should probably be moved from the `git` plugin to the core `lib/git.zsh` to remove this plugin dependency. Possibly, it should be renamed (e.g. to `git_current_branch`) to make it clear that it's for Git and not Mercurial or Bazaar or whatever, and a `current_branch` back-compatibility wrapper left in the `git` plugin.

Caused #4059.
Caused #4084.
Probably caused #2193.
Caused #331.
Caused by #329, in part.
Caused by #2723 (in #3809).

Fixed by #2386, I think, but that PR has a lot of other stuff going on, too, and has been sitting for 18 months without getting merged. Should probably do a smaller, easier-to-review patch to fix this issue independently.

**Comment by ncanceill** [comment]
Ouch, this is bad! Has it been around forever?

Anyway, I agree on the diagnosis and the suggested fix.

**Comment by apjanke** [comment]
Since May 2011, when #329 introduced the first `$(current_branch)` call in `git_prompt_ahead()`. The other `$(current_branch)` references seem to have shown up recently with PR #3809, which pulled in #2723 (2d40cc0bb315a92368e35c15c92f955bd901c7c3).

The `current_branch()` function uses `$_omz_git_git_cmd` but otherwise has no dependencies on other things in the `git` plugin. To pull it in to core lib, we'd need to pull the `$_omz_git_git_cmd` definition up too, or have `current_branch` use the plain `git` command. I think using plain `git` makes sense, because that's how the rest of core `lib/git.zsh` works.

@Stibbons, you're listed as the other maintainer of the `git` plugin. Thoughts?

**Comment by apjanke** [comment]
I've made the changes on a new branch on my fork, https://github.com/apjanke/oh-my-zsh/tree/git-move-current_branch-to-core (commit https://github.com/apjanke/oh-my-zsh/commit/1a5f8fdb6d0b03713b7aca35eabb89e2fc13a7bc). Couldn't resist doing a little cleanup while in there, too.

* Moves `current_branch()` from `git` plugin to core `lib/git.zsh`, renaming to `git_current_branch()`
* Updates `lib/git.zsh` to use its new `git_current_branch`, removing the plugin dependency
* Updates all plugins using `current_branch` to use `git_current_branch()`, removing a plugin dependency, and maybe allowing `current_branch()` back-compatibility wrapper to be removed at some point
* `local`-izes all the variables in `lib/git.zsh` functions to avoid leaks
* Has `lib/git.zsh` use `[[ ... ]]` tests instead of `[ ... ]` everywhere
* Formats comments and `function foo()` lines to be uniform

I'm pretty sure the `local`-ization is safe. Those variables look like they're for the functions' internal use only, I grepped for references to them in the themes and didn't find any, and all the git prompt building is done with `$(...)` function output capture instead of variable substitution wherever I've seen it in OMZ.

While we're in here, the `git` plugin function named `current_repository` seems to output a list of remotes. Is that the right name for it?
